### PR TITLE
[wasm] trim System.Private.Uri.dll

### PR DESCRIPTION
**Pull request by pavelsavara**
In the legacy interop, we have edge case use-case: when somebody would `bind_static_method("Foo")` where `void Foo(Uri uri)` the old marshaler would marshal string as `Uri` instance in Net6. 
This feature costs about 103KB of `System.Private.Uri.dll` uncompressed for all runtime/wasm users, regardles if they use it or not.

This changes the code so that ILLink would not protect Uri class and it could be trimmed out.

I think that if the user code calls methods with `Uri` parameter, they probably also have the `Uri` constructor in scope.

Even if they don't have it, they could configure ILLink to protect `Uri` after they enabled trimming.

My.csproj
```
  <ItemGroup>
    <TrimmerRootDescriptor Include="$(MSBuildThisFileDirectory)ILLink.Descriptors.xml" />
  </ItemGroup>
```

ILLink.Descriptors.xml
```
<linker>
  <assembly fullname="System.Private.Uri">
    <type fullname="System.Uri">
        <method signature="System.Void .ctor(System.String)" />
    </type>
  </assembly>
</linker>
```

### CODE REVIEW

**Review by radekdoulik — APPROVED** [review]
Given that it is a corner case, I think it is fine. For NET8 it will make sense to add a new feature switch for the legacy JS interop as discussed on discord and hide the reflection behind it.

File: src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Interop/LegacyExports.cs [hunk 1]
```diff
@@ -210,13 +211,24 @@ public static void CreateDateTimeRef(double ticks, out object result)
...
-            // if the user code has methods with Uri signature, this should work too
-            // System.Private.Uri is large assembly so it's worth trimming
-            var uriType = Type.GetType("System.Uri, System.Private.Uri");
+            if (uriType == null)
+            {
+                // StringBuilder to confuse ILLink, which is too smart otherwise
+                StringBuilder sb = new StringBuilder("System.Uri, System.Private.Uri");
+                uriType = Type.GetType(sb.ToString());
+            }
             if (uriType == null) throw new InvalidProgramException();
```
Comment (maraf) [inline src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Interop/LegacyExports.cs:232]: Can we explain here what happend and what to do to fix it?
Will it make sense to use `ValueStringBuilder` to skip even the single allocation here?

Comment (pavelsavara) [reply src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Interop/LegacyExports.cs]: Added better exception message and sample configuration for not trimming it.

**Review by maraf — APPROVED** [review]

**Comment by pavelsavara** [comment]
```
Uncaught Error: System.MissingMethodException: Constructor on type 'System.Uri' not found. Please consider to protect it's constructor from trimming.
 ---> System.MissingMethodException: Constructor on type 'System.Uri' not found.
   at System.RuntimeType.CreateInstanceImpl(BindingFlags , Binder , Object[] , CultureInfo )
   at System.Activator.CreateInstance(Type , BindingFlags , Binder , Object[] , CultureInfo , Object[] )
   at System.Activator.CreateInstance(Type , Object[] )
   at System.Runtime.InteropServices.JavaScript.LegacyExports.CreateUriRef(String uri, Object& result)
   --- End of inner exception stack trace ---
   at System.Runtime.InteropServices.JavaScript.LegacyExports.CreateUriRef(String uri, Object& result)
```

This is how it looks like when the constructor was trimmed.
I also tested that the ILink configuration makes it work again, see top post.

**Pull request merged**